File: fsharp/affine-cipher/AffineCipher.fs
module AffineCipher

open System

let cands = [1..2..11] @ [15..2..25]

let isNotValid (a: int) = 
    not (List.contains (a%26) cands)

let modInverse26 (a:int) =
    cands |> List.filter (fun i -> ((a%26) * (i%26)) % 26 = 1) |> List.exactlyOne

let letterToInt (c: char) =
    int (Char.ToLower(c)) - int 'a'

let rec intToLetter (i: int) =
    if i < 0 then 
        intToLetter (i + 26)
    else 
        char ((i%26) + int 'a')

let decode a b cipheredText = 
    if isNotValid a then 
        raise <| System.ArgumentException()
    else
        let affineDecode (c:char) =
            let modInv = modInverse26 a
            let y = letterToInt c
            intToLetter (modInv * (y - b))
        cipheredText 
        |> Seq.filter (Char.IsLetterOrDigit) 
        |> Seq.map (fun c -> if Char.IsDigit(c) then c else affineDecode c)
        |> Seq.toArray
        |> System.String


let encode a b plainText = 
    if isNotValid a then 
        raise <| System.ArgumentException()
    else
        let affineEncode (c:char) =
            let x = letterToInt c
            intToLetter (a * x + b)
        let listOf5s = 
            plainText 
            |> Seq.filter (Char.IsLetterOrDigit) 
            |> Seq.map (fun c -> if Char.IsDigit(c) then c else affineEncode c)
            |> Seq.chunkBySize 5 
            |> Seq.map (fun arr -> String(arr)) |> Seq.toList
        String.Join(" ", listOf5s)

File: csharp/affine-cipher/AffineCipher.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class AffineCipher
{
    static List<int> Cands = new List<int> { 1, 3, 5, 7, 9, 11, 15, 17, 19, 21, 23, 25 };
    static bool IsNotValid(int a) => !Cands.Contains(a % 26);
    static int ModInverse26(int a) => Cands.Where(c => ((a % 26) * (c % 26)) % 26 == 1).First();
    static int LetterToInt(char c) => char.ToLower(c) - 'a';
    static char IntToLetter(int i)
    {
        if (i < 0) return IntToLetter(i + 26);
        else return (char)((i % 26) + 'a');
    }
    public static string Encode(string plainText, int a, int b)
    {
        if (IsNotValid(a)) throw new ArgumentException("a not valid");
        char AffineEncode(char c)
        {
            int x = LetterToInt(c);
            return IntToLetter(a * x + b);
        }
        int helper = 0;
         var listOf5s = plainText
                        .Where(char.IsLetterOrDigit)
                        .Select(c => char.IsDigit(c) ? c : AffineEncode(c))
                        .GroupBy(x => helper++ / 5).Select(s => new string(s.ToArray())).ToArray();
        return string.Join(' ', listOf5s);
    }

    public static string Decode(string cipheredText, int a, int b)
    {
        if (IsNotValid(a)) throw new ArgumentException("a not valid");
        char AffineDecode(char c)
        {
            int modInv = ModInverse26(a);
            int y = LetterToInt(c);
            return IntToLetter(modInv * (y - b));
        }
        return new(cipheredText
                    .Where(char.IsLetterOrDigit)
                    .Select(c => char.IsDigit(c) ? c : AffineDecode(c))
                    .ToArray()
                   );
    }
}
